torch/csrc/api/include/torch/nn/modules/any.h

```diff
@@ -0,0 +1,215 @@
+#pragma once
+
+#include <torch/detail/any.h>
+
+#include <torch/csrc/autograd/variable.h>
+#include <torch/csrc/utils/memory.h>
+#include <torch/csrc/utils/variadic.h>
+
+#include <memory>
+#include <type_traits>
+#include <typeinfo>
+#include <utility>
+#include <vector>
+
+namespace torch {
+namespace detail {
+template <typename Head = void, typename... Tail>
+void check_not_lvalue_references() {
+  static_assert(
+      !std::is_lvalue_reference<Head>::value ||
+          std::is_const<typename std::remove_reference<Head>::type>::value,
+      "Modules stored inside AnyModule must not take (lvalue) references. Use pointers instead.");
+  check_not_lvalue_references<Tail...>();
+}
+
+template <>
+inline void check_not_lvalue_references<void>() {}
+} // namespace detail
+
+namespace nn {
+
+/// A class to store a type erasd module, whose `forward()` method can be
+/// invoked, with dynamic type checking.
+class AnyModule {
+ public:
+  /// Constructs an `AnyModule` from a `shared_ptr` to concrete module object.
+  template <typename ModuleType>
+  explicit AnyModule(std::shared_ptr<ModuleType> module)
+      : content_(make_holder(
+            std::move(module),
+            &std::remove_reference<ModuleType>::type::forward)) {}
+
+  /// Constructs an `AnyModule` from a concrete module object.
+  template <typename ModuleType>
+  explicit AnyModule(ModuleType&& module)
+      : AnyModule(
+            std::make_shared<ModuleType>(std::forward<ModuleType>(module))) {}
+
+  /// Move construction and assignment is allowed, and follows the default
+  /// behavior of move for `std::unique_ptr`.
+  AnyModule(AnyModule&&) = default;
+  AnyModule& operator=(AnyModule&&) = default;
+
+  /// Copy is disallowed.
+  AnyModule(const AnyModule& other) = delete;
+  AnyModule& operator=(const AnyModule& other) = delete;
+
+  /// Invokes `forward()` on the contained module with the given arguments, and
+  /// returns the return value as an `Any`. Use this method when chaining
+  /// `AnyModule`s in a loop.
+  template <
+      typename... ArgumentTypes,
+      typename = disable_for_t<Any, ArgumentTypes...>>
+  Any forward(ArgumentTypes&&... arguments) {
+    return content_->forward(std::forward<ArgumentTypes>(arguments)...);
+  }
+
+  /// A type erased `forward()` that invokes the actual `forward()` with a value
+  /// of any type, and returns a type erased object.
+  Any forward(Any&& any) {
+    std::vector<Any> vector;
+    vector.push_back(std::move(any));
+    return content_->forward(std::move(vector));
+  }
+
+  /// Attempts to cast the underlying module to the given module type. Throws an
+  /// exception if the types do not match.
+  template <typename T, typename = enable_if_module<T>>
+  T& get() {
```

Comment:
This is inconsistent to your `Any` implementation which returns by value. If you decide to change this here instead of the one in Any, you probably have to change the call site in Sequential so it doesn't suddenly copy.

Reply:
`Any` is an implementation detail and is used to store temporaries, while `AnyModule` is used to store modules permanently

Reply:
This restriction on `Any` isn't obvious from the code. Can we make `Any` generic enough to handle temporaries and permanent values? If yes, then it could be reused here.